Cannot ignore staff members

dagjaneiro opened the pull request:
## What does this PR do?

It prevents users from ignoring staff members.

## How do I test this PR?

- try to ignore a comment from a staff member
- A message is displayed: "Cannot ignore Staff members."

Addresses #1023.

dagjaneiro committed:
Add cannot ignore staff error

dagjaneiro committed:
Implement cannot ignore staff in users service

`UsersService.ignoreUsers` now checks if any user provided is a staff
member. In case it is it throws a ErrCannotIgnoreStaff error.

dagjaneiro committed:
Handle IgnoreUser mutation errors

We now check if the IgnoreUser mutation returned an error before
updating the apollo store. If the mutation returns an error we keep
the store as it is, otherwise the user is added to the ignore list.

kgardnr commented:
Thank you @dagjaneiro! We will review this shortly.

kgardnr committed:
Merge branch 'master' into cannot-ignore-staff

alinhle commented on plugins/talk-plugin-ignore-user/client/index.js:
```diff
@@ -14,7 +15,11 @@ export default {
       updateQueries: {
-        CoralEmbedStream_Embed: (previousData) => {
+        CoralEmbedStream_Embed: (previousData, {mutationResult: {data: {ignoreUser}}}) => {
+          if (get(ignoreUser, 'errors.length')) {
```
You can remove this part, our framework does not run this function when there was an error.

dagjaneiro replied:
Hi @cvle, thanks for your feedback. 

Actually, this was a source of confusion for me. 
It seems the framework only throws a GraphQL error if the error is a generic one. 
If you use an API error as a response, it wraps the error in a node specific to the mutation and prevents GraphQL from throwing an exception. 

I believe [this code](https://github.com/coralproject/talk/blob/master/graph/errorHandler.js#L8) is responsible for the behavior I just described.

The payload I'm getting in the frontend is:
```
{
  "data": {
    "ignoreUser": {
      "errors": [{
        "translation_key": "CANNOT_IGNORE_STAFF",
        "__typename": "GenericUserError"
      }],
      "__typename": "IgnoreUserResponse"
    }
  }
}
```
With this payload apollo is returning a successful `APOLLO_MUTATION_RESULT` instead of the expected `APOLLO_MUTATION_ERROR` which would prevent `updateQueries` from running.

Am I missing something to make this work without having to handle the error case explicitly as I did above? Thanks for your help 🙏

alinhle replied:
@dagjaneiro This is correct from the server side view. On the client the mutations runs through our framework that powers the plugin system and it detects the error specific to the mutation and doesnt run the updaters https://github.com/coralproject/talk/blob/master/client/coral-framework/hocs/withMutation.js#L111. It also fails the promise when it detects these types of errors here: https://github.com/coralproject/talk/blob/master/client/coral-framework/hocs/withMutation.js#L144

dagjaneiro replied:
@cvle with your guidance I figured out what's happening. 

When the `withMutation` hoc is processing the `updateQueries`, the [config array](https://github.com/coralproject/talk/blob/master/client/coral-framework/hocs/withMutation.js#L99) has two objects with keys `CoralEmbedStream_Embed`:
- the [featured-comments plugin ignoreUser mutation](https://github.com/coralproject/talk/blob/master/plugins/talk-plugin-featured-comments/client/index.js#L27)
- the [ignore-user plugin ignoreUser mutation I changed](https://github.com/coralproject/talk/pull/1078/files#diff-df3b0f249d219aa83e4324f0ed4fd808R16)

Because of that, when the [reducer function](https://github.com/coralproject/talk/blob/master/client/coral-framework/hocs/withMutation.js#L105) processes my mutation, the key is already present in the result object, so [the callback associated with the mutation](https://github.com/coralproject/talk/blob/master/client/coral-framework/hocs/withMutation.js#L120) doesn't have the error handling bailout code and ends up running `updateQueries` when it shouldn't.

This looks like a bug. If you're ok with it I can address it in this PR, unless there's a reason for it to be this way?

alinhle replied:
Thanks for your investigation @dagjaneiro. I will take a look too.

alinhle replied:
@dagjaneiro You are right, that something is odd, I will dive deeper tomorrow.

alinhle replied:
Thanks a ton!

alinhle reviewed (changes requested):
See remarks

kgardnr committed:
Merge branch 'master' into cannot-ignore-staff

kgardnr commented:
Hi @dagjaneiro we have some feedback for you - once it's addressed, this will be merged. Thank you again!

wyattjoh committed:
Merge branch 'master' into cannot-ignore-staff

kgardnr committed:
Merge branch 'master' into cannot-ignore-staff

kgardnr committed:
Merge branch 'master' into cannot-ignore-staff

wyattjoh committed:
Merge branch 'master' into cannot-ignore-staff

dagjaneiro committed:
Remove error handling code

Remove error handling code since the withMutation HoC already handles
errors for us.

dagjaneiro commented:
@cvle I've removed the error handling code from the `ignoreUser` mutation, once the issue with the `withMutation` HoC is resolved this should be good to merge.

alinhle commented:
@dagjaneiro thanks, pending PR is here https://github.com/coralproject/talk/pull/1100. After that we can merge yours.

alinhle committed:
Merge branch 'master' into cannot-ignore-staff
